R/count.R

```diff
@@ -139,27 +139,51 @@ process_single_count_target <- function(x) {
       # Used to temporarily check formats
       if (is.null(format_strings)) tmp_fmt <- gather_defaults.count_layer(current_env())
       if (count_missings && !(is.null(denom_ignore) || length(denom_ignore) == 0) &&
-         (("pct" %in% total_count_format$vars || "distinct_pct" %in% total_count_format$vars) ||
-         # Logic if no total_count format
-         (is.null(total_count_format) && is.null(format_strings) && ("pct" %in% tmp_fmt$n_counts$vars || "distinct_pct" %in% tmp_fmt$n_counts$vars)) ||
-         (is.null(total_count_format) && ("pct" %in% count_fmt$n_counts$vars || "distinct_pct" %in% count_fmt$n_counts$vars))
-         )
-         ) {
-           warning("Your total row is ignoring certain values. The 'pct' in this row may not be 100%",
-                   immediate. = TRUE)
-         }
+          (("pct" %in% total_count_format$vars || "distinct_pct" %in% total_count_format$vars) ||
+           # Logic if no total_count format
+           (is.null(total_count_format) && is.null(format_strings) && ("pct" %in% tmp_fmt$n_counts$vars || "distinct_pct" %in% tmp_fmt$n_counts$vars)) ||
+           (is.null(total_count_format) && ("pct" %in% count_fmt$n_counts$vars || "distinct_pct" %in% count_fmt$n_counts$vars))
+          )
+      ) {
+        warning("Your total row is ignoring certain values. The 'pct' in this row may not be 100%",
+                immediate. = TRUE)
+      }
     }
 
     if (is.null(count_row_prefix)) count_row_prefix <- ""
 
-    if (is.null(denoms_by)) denoms_by <- c(treat_var, cols)
+    # If a denoms variable is factor then it should be character for the denoms calculations
+    denoms_df_prep <- denoms_df %>%
+      mutate(
+        across(dplyr::where(is.factor), ~as.character(.))
+      )
+
+    # Nested counts might have summary var come through as numeric
+    if ('summary_var' %in% map_chr(denoms_by, as_name) && is.numeric(denoms_df_prep$summary_var)) {
+      denoms_df_prep$summary_var <- as.character(denoms_df_prep$summary_var)
+    }
+
+    # But if a summary_stat variable is factor, then the denoms needs to match this
+    # This happens if sorting was triggered for the variable as a factor
+    fct_cols <- map_lgl(summary_stat, is.factor)
+
+    if (any(fct_cols)) {
+      fct_cols_ns <- paste(names(fct_cols[fct_cols]), collapse="|")
+
+      denoms_df_prep <- denoms_df_prep %>%
+        mutate(
+          across(matches(fct_cols_ns), ~ factor(., levels=levels(summary_stat[[cur_column()]])))
+        )
```

Comment:
This is tricky to parse, can you add some comments surrounding what fct_cols_ns represents? the fct_cols[fct_cols] thing is breaking my brain

Reply:
I added comments. The gist is that factor levels fall off somewhere and need to be reset - and this happens based on sort settings. So from the summary stat data, I grab the variable that are factors and pull out there names, and then build the regex so I can use `across()` to loop across them and reset them back to factors with the original levels. This impacts the denoms later because 0 filling wasn't working otherwise. 

This was another challenging part of this entire section - variables flip between character and factor, or even numeric and factor, and it had to reconcile before it went into the denom totals. I think it's just part of the evolution of how many corner cases would pop up over time and it kind of naturally needs a refactor.

Reply:
I see now. Yep those comments really help. Looks good